File: apps/commerce-api/src/main/java/com/loopers/domain/commerceevent/CommerceEvent.java
package com.loopers.domain.commerceevent;

import java.time.LocalDateTime;
import java.util.Map;

import com.loopers.support.JsonMapConverter;

import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(name = "commerce_events")
public class CommerceEvent {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "event_id", unique = true, nullable = false)
    private String eventId;

    @Enumerated(EnumType.STRING)
    @Column(name = "event_type", nullable = false)
    private EventType eventType;

    @Column(name = "aggregate_id")
    private String aggregateId;

    @Column(name = "payload", columnDefinition = "TEXT")
    @Convert(converter = JsonMapConverter.class)
    private Map<String, Object> payload;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private CommerceEventStatus status = CommerceEventStatus.PENDING;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "retry_count")
    private Integer retryCount = 0;

    public CommerceEvent(String eventId, EventType eventType, String aggregateId,
        Map<String, Object> payload) {
        this.eventId = eventId;
        this.eventType = eventType;
        this.aggregateId = aggregateId;
        this.payload = payload;
        this.createdAt = LocalDateTime.now();
    }

    public static CommerceEvent from(CommerceEventCommand.Record command) {
        return new CommerceEvent(
            command.eventId(),
            command.eventType(),
            command.aggregateId(),
            command.payload());
    }
}

File: apps/commerce-streamer/src/main/kotlin/com/loopers/domain/commerceevent/CommerceEvent.kt
package com.loopers.domain.commerceevent

import com.loopers.support.JsonMapConverter
import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "commerce_events")
data class CommerceEvent(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(name = "event_id", unique = true, nullable = false)
    val eventId: String,

    @Column(name = "event_type", nullable = false)
    val eventType: String,

    @Column(name = "aggregate_id")
    val aggregateId: String,

    @Column(name = "payload", columnDefinition = "TEXT")
    @Convert(converter = JsonMapConverter::class)
    val payload: Map<String, Any>,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    val status: CommerceEventStatus = CommerceEventStatus.PENDING,

    @Column(name = "created_at", nullable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "retry_count")
    val retryCount: Int = 0,
) {
    fun processed(): Boolean {
        return status == CommerceEventStatus.PUBLISHED
    }
}

enum class CommerceEventStatus {
    PENDING,
    PUBLISHED,
    FAILED
}
